sw/device/tests/sim_dv/lc_ctrl_transition_impl.c

```diff
@@ -0,0 +1,119 @@
+  CHECK(lc_count == exp_lc_count,
+        "LC_count error, expected %0d but actual count is %0d", exp_lc_count,
+        lc_count);
+}
+
+/**
+ * Tests the state transition request handshake between LC_CTRL and OTP_CTRL.
+ *
+ * 1). OTP pre-load image with lc_count = `8`.
+ * 2). Backdoor write OTP's LC parition to `TestLocked1` state, and backdoor
+ * write OTP's `test_exit` token and `test_unlock` token to match the rand
+ * patterns.
+ * 3). `TestLocked1` state disabled CPU, so external testbench will drive JTAG
+ * interface to transit to `TestUnlocked2` state and increment the LC_CNT.
+ * 4). When LC_CTRL is ready, check LC_CNT and LC_STATE register.
+ * 5). Program LC state transition request to advance to `Prod` state.
+ * 6). Issue hard reset.
+ * 7). Wait for LC_CTRL is ready, then check if LC_STATE advanced to `Dev`
+ * state, and lc_count advanced to `9`.
+ * 8). Issue hard reset and override OTP's LC partition, and reset LC state to
+ * `TestUnlocked2` state.
+ */
+
+bool execute_lc_ctrl_transition_test(bool use_ext_clk) {
+  LOG_INFO("Start LC_CTRL transition test.");
+
+  mmio_region_t lc_reg = mmio_region_from_addr(TOP_EARLGREY_LC_CTRL_BASE_ADDR);
+  CHECK_DIF_OK(dif_lc_ctrl_init(lc_reg, &lc));
+
+  LOG_INFO("Read and check LC state.");
+  dif_lc_ctrl_state_t curr_state;
+  CHECK_DIF_OK(dif_lc_ctrl_get_state(&lc, &curr_state));
+
+  // The OTP preload image hardcodes the initial LC state transition count to 8.
+  // With each iteration of test, there are two LC_CTRL state transitions.
+  // And the first LC_CTRL state transition is done via external JTAG interface.
+  // `kTestIterationCount` starts with 1 in SystemVerilog.
+  const uint8_t kLcStateTransitionCount = 8 + 1 + (kTestIterationCount - 1) * 2;
+
+  if (curr_state == kDifLcCtrlStateTestUnlocked2) {
+    // LC TestUnlocked2 is the intial test state for this sequence.
+    // The sequence will check if lc_count matches the preload value.
+    check_lc_state_transition_count(kLcStateTransitionCount);
+
+    // Request lc_state transfer to Dev state.
+    dif_lc_ctrl_token_t token;
+    dif_lc_ctrl_settings_t settings;
+    settings.clock_select =
+        use_ext_clk ? kDifLcCtrlExternalClockEn : kDifLcCtrlInternalClockEn;
+    for (int i = 0; i < LC_TOKEN_SIZE; i++) {
+      token.data[i] = kLcExitToken[i];
+    }
+    CHECK(dif_lc_ctrl_mutex_try_acquire(&lc) == kDifOk);
+    CHECK(dif_lc_ctrl_transition(&lc, kDifLcCtrlStateDev, &token, &settings) ==
+              kDifOk,
+          "LC_transition failed!");
+
+    LOG_INFO("Waiting for LC transtition done and reboot.");
+    wait_for_interrupt();
+
+  } else {
+    // LC Dev is the requested state from previous lc_state transition request.
+    // Once the sequence checks current state and count via CSRs, the test can
+    // exit successfully.
+    CHECK(curr_state == kDifLcCtrlStateDev, "State transition failed!");
+    check_lc_state_transition_count(kLcStateTransitionCount + 1);
+    return true;
```

Comment:
when use_ext_clk = 1, seems like we don't really know if the clock switch happens or not. If it's still using internal clock, we actually don't know.

Reply:
True. We could configure the external clock frequency to be 96 MHz and confirm by inspection that the test takes less time, or via waves. But it seems there is no other way to confirm this. I'll see if someone comes up with a suggestion from the chat.